packages/SwingSet/src/kernel/virtualObjectManager.js

```diff
@@ -0,0 +1,342 @@
+   *
+   * @param {string} instanceKey  The instance ID of the object whose state is
+   *    being stored.
+   * @param {*} rawData  A data object representing the state to be written.
+   */
+  function store(instanceKey, rawData) {
+    syscall.vatstoreSet(instanceKey, JSON.stringify(rawData));
+  }
+
+  /**
+   * Make a new kind of virtual object.
+   *
+   * @param {*} instanceMaker  A function of the form `instanceMaker(state)` that
+   *    will return a representative instance wrapped around the given state.
+   *
+   * @returns {*} a maker function that can be called to manufacture new
+   *    instance of this kind of object.  The parameters of the maker function
+   *    are those of the `initialize` method implemented in the representative
+   *    produced by the `instanceMaker` parameter.
+   */
+  function makeKind(instanceMaker) {
+    const kindID = `${allocateExportID()}`;
+    let nextInstanceID = 1;
+
+    function makeInnerSelf(instanceKey, rawData) {
+      const innerSelf = {
+        prev: undefined,
+        next: undefined,
+        instanceKey,
+        rawData,
+        dropData() {
+          store(instanceKey, innerSelf.rawData);
+          innerSelf.rawData = null;
+        },
+      };
+      return innerSelf;
+    }
+
+    function makeRepresentative(innerSelf, initArgs) {
+      function ensureState() {
+        if (!innerSelf.rawData) {
+          const instanceKey = innerSelf.instanceKey;
+          innerSelf = cache.lookup(instanceKey);
+          if (!innerSelf) {
+            innerSelf = makeInnerSelf(instanceKey, fetch(instanceKey));
+            cache.remember(innerSelf);
+          }
+        }
+        cache.refresh(innerSelf);
+      }
+
+      function wrapData() {
+        innerSelf.activeData = {};
+        for (const prop of Object.getOwnPropertyNames(innerSelf.rawData)) {
+          Object.defineProperty(innerSelf.activeData, prop, {
+            get: () => {
+              ensureState();
+              return m.unserialize(innerSelf.rawData[prop]);
+            },
+            set: value => {
+              ensureState();
+              innerSelf.rawData[prop] = m.serialize(value);
+            },
+          });
+        }
+        return harden(innerSelf.activeData);
+      }
+
+      if (!innerSelf.rawData && initArgs) {
+        innerSelf.rawData = {};
+        const tempInstance = instanceMaker(innerSelf.rawData);
+        tempInstance.initialize(...initArgs);
+        for (const prop of Object.getOwnPropertyNames(innerSelf.rawData)) {
+          innerSelf.rawData[prop] = m.serialize(innerSelf.rawData[prop]);
+        }
+      }
+      if (!innerSelf.activeData) {
+        wrapData();
+      }
+      const rawRepresentative = instanceMaker(innerSelf.activeData);
+      delete rawRepresentative.initialize;
+      const representative = harden(rawRepresentative);
+      cache.remember(innerSelf);
+      valToSlotTable.set(representative, innerSelf.instanceKey);
```

Comment:
I suspect this could be simpler if `valToSlotTable` were the responsibility of the caller, but I haven't thought it through completely.

Reply:
I totally don't follow this.

Reply:
Hm. `valToSlotTable` is used by two things: a `set` when we create a new representative, and a `get` when someone does a lookup in the weak-store (to distingusih between keys that are representatives, and keys which are plain objects).  And then liveslots needs it for its usual purpose. Eh, yeah, I'm probably wrong, nevermind.